File: matching-engine/cpp/src/reference_trace.cpp
// reference_trace.cpp — replay reference_scenario.txt through the C++ engine.
//
// Prints trades and the final book in exactly the format reference_trace.py
// uses, so diffing the two outputs proves the ports agree.

#include "order_book.hpp"

#include <filesystem>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <string>

namespace fs = std::filesystem;
using namespace engine;

namespace {

std::string fixed2(Price p) {
    std::ostringstream out;
    out << std::fixed << std::setprecision(2) << p;
    return out.str();
}

fs::path default_scenario() {
    // Walk up from the binary to find the scenario next to the Python sources.
    for (fs::path dir = fs::current_path(); !dir.empty(); dir = dir.parent_path()) {
        const auto candidate = dir / "reference_scenario.txt";
        if (fs::exists(candidate)) return candidate;
        if (!dir.has_relative_path()) break;
    }
    return "reference_scenario.txt";
}

}  // namespace

int main(int argc, char** argv) {
    const fs::path scenario = argc > 1 ? fs::path(argv[1]) : default_scenario();

    std::ifstream in(scenario);
    if (!in) {
        std::cerr << "Error: could not open scenario '" << scenario.string() << "'.\n";
        return 1;
    }

    OrderBook book;
    OrderId next_id = 1;
    std::string line;

    while (std::getline(in, line)) {
        // Strip comments and blanks.
        const auto hash = line.find('#');
        if (hash != std::string::npos) line = line.substr(0, hash);
        std::istringstream fields(line);
        std::string kind;
        if (!(fields >> kind)) continue;

        if (kind == "C") {
            OrderId id{};
            fields >> id;
            std::cout << "CANCEL " << id << (book.cancel(id) ? " ok" : " miss") << "\n";
            continue;
        }

        std::string side_text, price_text;
        Quantity qty{};
        fields >> side_text >> price_text >> qty;

        Order order;
        order.id = next_id++;
        order.side = side_text == "B" ? Side::Buy : Side::Sell;
        order.type = kind == "L" ? OrderType::Limit : OrderType::Market;
        order.quantity = qty;
        if (price_text != "-") order.price = std::stod(price_text);

        std::cout << "SUBMIT " << order.id << " "
                  << (order.side == Side::Buy ? "BUY" : "SELL") << " "
                  << (order.type == OrderType::Limit ? "LIMIT" : "MARKET") << " "
                  << (order.price ? fixed2(*order.price) : "-") << " " << qty << "\n";

        for (const auto& trade : book.submit(order)) {
            std::cout << "  TRADE maker=" << trade.maker_order_id
                      << " taker=" << trade.taker_order_id
                      << " price=" << fixed2(trade.price)
                      << " qty=" << trade.quantity << "\n";
        }
    }

    std::cout << "BOOK\n";
    const auto depth = book.depth(10);
    for (const auto& entry : depth.bids)
        std::cout << "  BID " << fixed2(entry.price) << " " << entry.quantity << "\n";
    for (const auto& entry : depth.asks)
        std::cout << "  ASK " << fixed2(entry.price) << " " << entry.quantity << "\n";

    const auto bid = book.best_bid();
    const auto ask = book.best_ask();
    const auto spread = book.spread();
    std::cout << "BEST_BID " << (bid ? fixed2(*bid) : "-") << "\n";
    std::cout << "BEST_ASK " << (ask ? fixed2(*ask) : "-") << "\n";
    std::cout << "SPREAD " << (spread ? fixed2(*spread) : "-") << "\n";
    std::cout << "RESTING " << book.resting_count() << "\n";

    return 0;
}

File: matching-engine/reference_trace.py
"""reference_trace.py — replay reference_scenario.txt through the Python engine.

Prints every trade and the final book in a fixed format. The C++ engine has a
matching runner; diffing the two outputs is what proves the ports agree.

Usage:
    python reference_trace.py [scenario.txt]
"""

import os
import sys

from book import OrderBook
from order import Order, OrderType, Side


def parse_scenario(path):
    """Yield ('submit', side, type, price, qty) or ('cancel', order_id)."""
    with open(path, "r", encoding="utf-8") as fh:
        for raw in fh:
            line = raw.strip()
            if not line or line.startswith("#"):
                continue
            parts = line.split()
            if parts[0] == "C":
                yield ("cancel", int(parts[1]))
            else:
                order_type = OrderType.LIMIT if parts[0] == "L" else OrderType.MARKET
                side = Side.BUY if parts[1] == "B" else Side.SELL
                price = None if parts[2] == "-" else float(parts[2])
                yield ("submit", side, order_type, price, int(parts[3]))


def main():
    scenario = sys.argv[1] if len(sys.argv) > 1 else os.path.join(
        os.path.dirname(os.path.abspath(__file__)), "reference_scenario.txt")

    book = OrderBook()
    next_id = 1

    for op in parse_scenario(scenario):
        if op[0] == "cancel":
            _, order_id = op
            removed = book.cancel(order_id)
            print(f"CANCEL {order_id} {'ok' if removed else 'miss'}")
            continue

        _, side, order_type, price, qty = op
        order = Order(side=side, quantity=qty, price=price,
                      order_type=order_type, order_id=next_id)
        next_id += 1

        price_text = "-" if price is None else f"{price:.2f}"
        print(f"SUBMIT {order.order_id} {side.value} {order_type.value} {price_text} {qty}")

        for trade in book.submit(order):
            print(f"  TRADE maker={trade.maker_order_id} taker={trade.taker_order_id} "
                  f"price={trade.price:.2f} qty={trade.quantity}")

    print("BOOK")
    depth = book.depth(levels=10)
    for price, quantity in depth["bids"]:
        print(f"  BID {price:.2f} {quantity}")
    for price, quantity in depth["asks"]:
        print(f"  ASK {price:.2f} {quantity}")

    bid, ask = book.best_bid(), book.best_ask()
    print(f"BEST_BID {'-' if bid is None else f'{bid:.2f}'}")
    print(f"BEST_ASK {'-' if ask is None else f'{ask:.2f}'}")
    spread = book.spread()
    print(f"SPREAD {'-' if spread is None else f'{spread:.2f}'}")
    print(f"RESTING {len(book.orders)}")


if __name__ == "__main__":
    main()
